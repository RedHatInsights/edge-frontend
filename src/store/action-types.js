import keyBy from 'lodash/keyBy';
import flatMap from 'lodash/flatMap';

export const LOAD_DEVICE_SUMMARY = 'LOAD_DEVICE_SUMMARY';
export const LOAD_ACTIVE_IMAGES = 'LOAD_ACTIVE_IMAGES';
export const LOAD_EDGE_IMAGES = 'LOAD_EDGE_IMAGES';
export const LOAD_EDGE_IMAGE_SETS = 'LOAD_EDGE_IMAGE_SETS';
export const LOAD_EDGE_IMAGE_PACKAGES = 'LOAD_EDGE_IMAGE_PACKAGES';
export const LOAD_GROUPS = 'LOAD_GROUPS';
export const LOAD_GROUP_DETAIL = 'LOAD_GROUP_DETAIL';
export const LOAD_TRESHOLD = 'LOAD_TRESHOLD';
export const LOAD_DEVICES_INFO = 'LOAD_DEVICES_INFO';
export const LOAD_CANARIES_INFO = 'LOAD_CANARIES_INFO';
export const LOAD_GROUP_DEVICES_INFO = 'LOAD_GROUP_DEVICES_INFO';
export const LOAD_IMAGE_STATUS = 'LOAD_IMAGE_STATUS';
export const LOAD_IMAGE_DETAIL = 'LOAD_IMAGE_DETAIL';
export const LOAD_IMAGE_SET_DETAIL = 'LOAD_IMAGE_SET_DETAIL';
export const CREATE_NEW_IMAGE = 'CREATE_NEW_IMAGE';
export const CREATE_NEW_IMAGE_RESET = `${CREATE_NEW_IMAGE}_RESET`;
export const POLLING_IMAGES = 'POLLING_IMAGES';
export const LOAD_DEVICE_TABLE = 'LOAD_DEVICE_TABLE';
export const UPDATE_DISPLAY_NAME = 'UPDATE_DISPLAY_NAME';
export const REMOVE_ENTITY = 'REMOVE_ENTITY';

const asyncActions = flatMap(
  [
    LOAD_EDGE_IMAGES,
    LOAD_EDGE_IMAGE_SETS,
    LOAD_EDGE_IMAGE_PACKAGES,
    LOAD_ACTIVE_IMAGES,
    LOAD_GROUPS,
    LOAD_GROUP_DETAIL,
    LOAD_TRESHOLD,
    LOAD_DEVICES_INFO,
    LOAD_CANARIES_INFO,
    LOAD_GROUP_DEVICES_INFO,
    LOAD_IMAGE_STATUS,
    LOAD_IMAGE_DETAIL,
    LOAD_IMAGE_SET_DETAIL,
    CREATE_NEW_IMAGE,
    LOAD_DEVICE_TABLE,
    UPDATE_DISPLAY_NAME,
    REMOVE_ENTITY,
  ],
  (a) => [a, `${a}_PENDING`, `${a}_FULFILLED`, `${a}_REJECTED`]
);
export const ACTION_TYPES = keyBy(asyncActions, (k) => k);

export const SELECT_ENTITY = 'SELECT_ENTITY';
export const PRE_SELECT_ENTITY = 'PRE_SELECT_ENTITY';
export const CLEAN_ENTITIES = 'CLEAN_ENTITIES';
