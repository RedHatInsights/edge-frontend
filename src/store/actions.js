import {
  ACTION_TYPES,
  LOAD_TRESHOLD,
  LOAD_DEVICES_INFO,
  LOAD_CANARIES_INFO,
  LOAD_GROUP_DETAIL,
  LOAD_GROUP_DEVICES_INFO,
  SELECT_ENTITY,
  PRE_SELECT_ENTITY,
  CLEAN_ENTITIES,
  LOAD_ACTIVE_IMAGES,
  LOAD_EDGE_IMAGES,
  LOAD_EDGE_IMAGE_SETS,
  LOAD_EDGE_IMAGE_PACKAGES,
  LOAD_DEVICE_SUMMARY,
  LOAD_IMAGE_STATUS,
  LOAD_IMAGE_DETAIL,
  LOAD_IMAGE_SET_DETAIL,
  CREATE_NEW_IMAGE,
  POLLING_IMAGES,
  LOAD_DEVICE_TABLE,
} from './action-types';

import {
  fetchImageStatus,
  getImageById,
  fetchEdgeImages,
  fetchEdgeImageSets,
  getImagePackageMetadata,
  createImage,
  getImageSet,
  fetchActiveImages,
} from '../api/images';
import { getInventory } from '../api/devices';
import { hosts } from '../api';
import { deleteSystemsById } from '../utils';

export const loadThreshold = () => ({
  type: LOAD_TRESHOLD,
  payload: () => {},
});

export const loadDevicesInfo = () => ({
  type: LOAD_DEVICES_INFO,
  payload: () => {},
});

export const loadCanariesInfo = () => ({
  type: LOAD_CANARIES_INFO,
  payload: () => {},
});

export const loadGroupsDetail = () => ({
  type: LOAD_GROUP_DETAIL,
  payload: () => {},
});

export const loadGroupDevicesInfo = () => ({
  type: LOAD_GROUP_DEVICES_INFO,
  payload: () => {},
});

export const selectEntity = (id, selected) => ({
  type: SELECT_ENTITY,
  payload: {
    id,
    selected,
  },
});

export const preSelectEntity = (id, selected) => ({
  type: PRE_SELECT_ENTITY,
  payload: {
    id,
    selected,
  },
});

export const cleanEntities = () => ({
  type: CLEAN_ENTITIES,
});

export const loadImages = (dispatch, pagination) => {
  dispatch({
    type: LOAD_ACTIVE_IMAGES,
    payload: fetchActiveImages(pagination),
    meta: {
      notifications: {
        rejected: {
          variant: 'danger',
          title: 'Cannot show images data',
          description: 'Failed receiving images from image-builder',
        },
      },
    },
    // the '.catch' part is necessary because redux-promise-middleware throws the error on REJECTED
    // and to avoid the app exploding I need to catch it here.
    // THANK you redux-promise-middleware for not allowing to customize this behavior. ðŸ˜ 
  }).catch(() => null);
};

export const loadDeviceSummary = (dispatch) => {
  dispatch({
    type: LOAD_DEVICE_SUMMARY,
    payload: () => {},
    meta: {
      notifications: {
        rejected: {
          variant: 'danger',
          title: 'Cannot show system summary data',
          description: 'Failed receiving system summary data from inventory',
        },
      },
    },
    // the '.catch' part is necessary because redux-promise-middleware throws the error on REJECTED
    // and to avoid the app exploding I need to catch it here.
    // THANK you redux-promise-middleware for not allowing to customize this behavior. ðŸ˜ 
  }).catch(() => null);
};

export const loadImageStatus = (dispatch, imageId) => {
  dispatch({
    type: LOAD_IMAGE_STATUS,
    payload: fetchImageStatus({ id: imageId }),
  }).catch(() => null);
};

export const loadImageDetail = (dispatch, imageId) => {
  dispatch({
    type: LOAD_IMAGE_DETAIL,
    payload: getImageById({ id: imageId }),
  }).catch(() => null);
};

export const createNewImage = (dispatch, payload, callback) => {
  dispatch({
    type: CREATE_NEW_IMAGE,
    payload: createImage(payload),
  })
    .then(callback)
    .catch(() => null);
};

export const loadEdgeImages = (dispatch, query) => {
  dispatch({
    type: LOAD_EDGE_IMAGES,
    payload: fetchEdgeImages(query),
  }).catch(() => null);
};

export const loadEdgeImageSets = (dispatch, query) => {
  dispatch({
    type: LOAD_EDGE_IMAGE_SETS,
    payload: fetchEdgeImageSets(query),
  }).catch(() => null);
};

export const setPolling = (toStart, interval) => {
  const subAction = toStart ? 'START' : 'END';
  const payload = toStart ? { interval } : {};
  return {
    type: `${POLLING_IMAGES}_${subAction}`,
    ...payload,
  };
};

export const addImageToPoll = ({ id, name }) => {
  return {
    type: `${POLLING_IMAGES}_ADD`,
    payload: {
      name,
      id,
    },
  };
};

export const removeImagesToPoll = (ids) => {
  return {
    type: `${POLLING_IMAGES}_REMOVE`,
    ids,
  };
};

export const loadImageSetDetail = (dispatch, urlParam, query) => {
  dispatch({
    type: LOAD_IMAGE_SET_DETAIL,
    payload: getImageSet({ id: urlParam, q: query }),
  }).catch(() => null);
};

export const loadImagePackageMetadata = (dispatch, imageId) => {
  dispatch({
    type: LOAD_EDGE_IMAGE_PACKAGES,
    payload: getImagePackageMetadata(imageId),
  }).catch(() => null);
};

export const loadDeviceTable = (dispatch) => {
  dispatch({
    type: LOAD_DEVICE_TABLE,
    payload: getInventory(),
  }).catch(() => null);
};

export const editDisplayName = (id, value, origValue) => ({
  type: ACTION_TYPES.UPDATE_DISPLAY_NAME,
  payload: hosts.apiHostPatchHostById([id], { display_name: value }),
  meta: {
    id,
    value,
    origValue,
    notifications: {
      fulfilled: {
        variant: 'success',
        title: `Display name for entity with ID ${id} has been changed to ${value}`,
        dismissable: true,
      },
    },
  },
});

export const deleteEntity = (systems, displayName) => ({
  type: ACTION_TYPES.REMOVE_ENTITY,
  payload: deleteSystemsById(systems),
  meta: {
    notifications: {
      fulfilled: {
        variant: 'success',
        title: 'Delete operation finished',
        description: `${displayName} has been successfully removed.`,
        dismissable: true,
      },
    },
    systems,
  },
});
