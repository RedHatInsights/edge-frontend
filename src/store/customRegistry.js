import ReducerRegistry from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

export const createRegistry = (
  initialState = {},
  middleware = [],
  composeEnhancersDefault
) => {
  return new ReducerRegistry(
    initialState,
    [...middleware],
    composeEnhancersDefault
  );
};
