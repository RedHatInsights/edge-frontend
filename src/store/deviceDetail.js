import GeneralInformationTab from '../components/DeviceDetail';
import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';

const entityLoaded = (state) => {
  return {
    ...state,
    loaded: true,
    activeApps: [
      {
        title: 'General information',
        name: 'general_information',
        component: GeneralInformationTab,
      },
    ],
  };
};

export const deviceDetail = applyReducerHash({
  LOAD_ENTITY_FULFILLED: entityLoaded,
});
