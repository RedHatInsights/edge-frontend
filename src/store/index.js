import { createContext } from 'react';
import { ReducerRegistry } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import { polling } from './polling';

export const RegistryContext = createContext({
  getRegistry: () => {},
});

export function init(...middleware) {
  return new ReducerRegistry({}, [
    promiseMiddleware,
    notificationsMiddleware({
      errorDescriptionKey: ['detail', 'stack'],
    }),
    polling(),
    ...middleware,
  ]);
}
