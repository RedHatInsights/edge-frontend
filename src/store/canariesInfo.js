import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { LOAD_CANARIES_INFO } from './action-types';

const initialState = {};

const loadCanariesInfoPending = (state) => ({
  ...state,
  canariesInfo: [],
  isLoading: true,
});
const loadCanariesInfoFulfilled = (state, { payload }) => ({
  ...state,
  canariesInfo: payload?.results || [],
  isLoading: false,
});

export default applyReducerHash(
  {
    [`${LOAD_CANARIES_INFO}_PENDING`]: loadCanariesInfoPending,
    [`${LOAD_CANARIES_INFO}_FULFILLED`]: loadCanariesInfoFulfilled,
  },
  initialState
);
