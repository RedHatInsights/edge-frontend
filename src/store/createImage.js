import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { CREATE_NEW_IMAGE } from './action-types';

const initialState = {};

const createNewImagePending = () => ({
  isLoading: true,
  hasError: false,
  error: null,
});

const createNewImageFulfilled = () => {
  return {
    isLoading: false,
    hasError: false,
    error: null,
  };
};

const createNewImageRejected = (_state, { payload }) => ({
  isLoading: false,
  hasError: true,
  error: payload,
});

const createNewImageReset = () => ({
  isLoading: false,
  hasError: false,
  error: null,
});

export default applyReducerHash(
  {
    [`${CREATE_NEW_IMAGE}_PENDING`]: createNewImagePending,
    [`${CREATE_NEW_IMAGE}_FULFILLED`]: createNewImageFulfilled,
    [`${CREATE_NEW_IMAGE}_REJECTED`]: createNewImageRejected,
    [`${CREATE_NEW_IMAGE}_RESET`]: createNewImageReset,
  },
  initialState
);
