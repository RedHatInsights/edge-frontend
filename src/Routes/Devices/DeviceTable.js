import React, { useEffect, useState, useContext, Suspense } from 'react';
import GeneralTable from '../../components/general-table/GeneralTable';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { routes as paths } from '../../../package.json';
import { Link } from 'react-router-dom';
import { DateFormat } from '@redhat-cloud-services/frontend-components/DateFormat';
import { cellWidth } from '@patternfly/react-table';
import { Label, Bullseye, Spinner } from '@patternfly/react-core';
import { shallowEqual, useSelector } from 'react-redux';
import { loadDeviceTable } from '../../store/actions';
import { RegistryContext } from '../../store';
import { deviceTableReducer } from '../../store/reducers';
import CustomEmptyState from '../../components/Empty';
import {
  CheckCircleIcon,
  ExclamationTriangleIcon,
  InProgressIcon,
} from '@patternfly/react-icons';

const UpdateDeviceModal = React.lazy(() =>
  import(/* webpackChunkName: "CreateImageWizard" */ './UpdateDeviceModal')
);

const defaultFilters = [
  {
    label: 'Name',
    type: 'text',
  },
  {
    label: 'Status',
    type: 'checkbox',
    options: [
      { option: 'Building', value: 'BUILDING' },
      { option: 'Created', value: 'CREATED' },
      { option: 'Error', value: 'ERROR' },
      { option: 'Ready', value: 'SUCCESS' },
    ],
  },
];

const columnNames = [
  {
    title: 'Name',
    type: 'name',
    sort: true,
    columnTransforms: [cellWidth(35)],
  },
  {
    title: 'Groups',
    type: 'groups',
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Last Seen',
    type: 'last_seen',
    sort: true,
    columnTransforms: [cellWidth(15)],
  },
  {
    title: 'Image',
    type: 'image',
    sort: false,
    columnTransforms: [cellWidth(20)],
  },
  {
    title: 'Status',
    type: 'status',
    sort: false,
    columnTransforms: [cellWidth(15)],
  },
];

const DeviceStatus = ({ Device }) => {
  const status = getDeviceStatus(Device);
  const statusType = {
    booting: (
      <Label
        className="pf-u-mt-sm"
        color="blue"
        icon={<InProgressIcon color="blue" />}
      >
        Booting
      </Label>
    ),
    running: (
      <Label
        className="pf-u-mt-sm"
        color="green"
        icon={<CheckCircleIcon color="green" />}
      >
        Running
      </Label>
    ),
    updateAvailable: (
      <Label
        className="pf-u-mt-sm"
        color="orange"
        icon={<ExclamationTriangleIcon />}
      >
        Update Available
      </Label>
    ),
  };

  return statusType[status];
};

const getDeviceStatus = (deviceData) =>
  deviceData?.ImageInfo?.UpdatesAvailable
    ? 'updateAvailable'
    : deviceData?.Device?.Booted
    ? 'running'
    : 'booting';

const createRows = (devices) =>
  devices?.map((device) => ({
    id: device?.Device?.UUID,
    display_name: device?.Device?.DeviceName,
    updateImageData: device?.ImageInfo?.UpdatesAvailable?.[0],
    deviceStatus: getDeviceStatus(device),
    noApiSortFilter: [
      device?.Device?.DeviceName,
      '',
      device?.Device?.LastSeen,
      device?.ImageInfo?.Image?.Name || '',
      getDeviceStatus(device),
    ],
    cells: [
      {
        title: (
          <Link to={`${paths['inventory']}/${device?.Device?.UUID}`}>
            {device?.Device?.DeviceName}
          </Link>
        ),
      },
      {
        title: '-',
      },
      {
        title: <DateFormat date={device?.Device?.LastSeen} />,
      },
      {
        title: (
          <Link
            to={`${paths['manage-images']}/${device?.ImageInfo?.Image?.ImageSetID}/versions/${device?.ImageInfo?.Image?.ID}/details`}
          >
            {device?.ImageInfo?.Image?.Name}
          </Link>
        ),
      },
      {
        title: <DeviceStatus Device={device} />,
      },
    ],
  }));

const DeviceTable = ({ skeletonRowQuantity }) => {
  const { getRegistry } = useContext(RegistryContext);
  const [reload, setReload] = useState([]);
  const dispatch = useDispatch();
  const [updateModal, setUpdateModal] = useState({
    isOpen: false,
    deviceData: null,
    imageData: null,
  });

  const { count, data, isLoading, hasError } = useSelector(
    ({ deviceTableReducer }) => ({
      count: deviceTableReducer?.data?.count || 0,
      data: deviceTableReducer?.data?.data || [],
      isLoading: deviceTableReducer?.isLoading,
      hasError: deviceTableReducer?.hasError,
    }),
    shallowEqual
  );

  useEffect(() => {
    const registered = getRegistry().register({ deviceTableReducer });
    loadDeviceTable(dispatch);
    return () => registered();
  }, [reload]);

  const actionResolver = () => {
    return [
      {
        title: 'Update Device',
        onClick: (_event, _rowId, rowData) => {
          setUpdateModal((prevState) => {
            return {
              ...prevState,
              isOpen: true,
              deviceData: {
                id: rowData?.id,
                display_name: rowData?.display_name,
              },
              imageData: rowData?.updateImageData,
            };
          });
        },
      },
    ];
  };

  const areActionsDisabled = (rowData) =>
    rowData?.deviceStatus !== 'updateAvailable';

  return (
    <>
      {!isLoading && !count > 0 ? (
        <CustomEmptyState
          data-testid="general-table-empty-state-no-data"
          icon={'plus'}
          title={'Connect edge devices'}
          body={
            'Connect and manage edge devices here after registering them via the console. To start, create a RHEL for Edge image and install it to your target device.'
          }
          secondaryActions={[
            {
              title: 'How to connect a device',
              link: '/',
              type: 'link',
            },
          ]}
        />
      ) : (
        <GeneralTable
          apiFilterSort={false}
          filters={defaultFilters}
          loadTableData={loadDeviceTable}
          tableData={{
            count: count,
            isLoading: isLoading,
            hasError: hasError,
          }}
          columnNames={columnNames}
          rows={createRows(data) || []}
          actionResolver={actionResolver}
          areActionsDisabled={areActionsDisabled}
          defaultSort={{ index: 2, direction: 'desc' }}
          // toolbarButtons={[
          //   {
          //     title: 'Group Selected',
          //     click: () => console.log('Group Selected'),
          //   },
          // ]}
          hasCheckbox={true}
          skeletonRowQuantity={skeletonRowQuantity}
        />
      )}
      {updateModal.isOpen && (
        <Suspense
          fallback={
            <Bullseye>
              <Spinner />
            </Bullseye>
          }
        >
          <UpdateDeviceModal
            navigateBack={() => {
              history.push({ pathname: history.location.pathname });
              setUpdateModal((prevState) => {
                return {
                  ...prevState,
                  isOpen: false,
                };
              });
            }}
            setUpdateModal={setUpdateModal}
            updateModal={updateModal}
            refreshTable={() => setReload(true)}
          />
        </Suspense>
      )}
    </>
  );
};
DeviceTable.propTypes = {
  imageData: PropTypes.object,
  urlParam: PropTypes.string,
  openUpdateWizard: PropTypes.func,
  skeletonRowQuantity: PropTypes.number,
};

export default DeviceTable;
