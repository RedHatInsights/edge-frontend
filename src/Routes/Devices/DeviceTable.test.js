import React from 'react';
import DeviceTable from './DeviceTable';
import { Provider } from 'react-redux';
import { init, RegistryContext } from '../../store';
import { render, screen } from '@testing-library/react';
import logger from 'redux-logger';
import { MemoryRouter } from 'react-router-dom';

describe('ImageSets table', () => {
  it('to render correctly', async () => {
    const registry = init(logger);

    const { container } = render(
      <RegistryContext.Provider
        value={{
          getRegistry: () => registry,
        }}
      >
        <Provider store={registry.getStore()}>
          <DeviceTable />
        </Provider>
      </RegistryContext.Provider>,
      { wrapper: MemoryRouter }
    );

    const headerArray = ['Name', 'Groups', 'Last Seen', 'Image', 'Status'];

    //expect(screen.getByLabelText('Select all rows').checked).toBeFalsy();
    expect(
      screen.getAllByRole('columnheader').forEach((header, i) => {
        expect(header.innerHTML === headerArray[i]).toBeTruthy();
      })
    );
    expect(screen.getByRole('button', { name: 'Options menu' })).toBeDefined();
    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
