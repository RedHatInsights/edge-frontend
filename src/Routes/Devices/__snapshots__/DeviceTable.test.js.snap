// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSets table to render correctly 1`] = `
<div
  class="pf-c-toolbar"
  data-ouia-component-id="OUIA-Generated-Toolbar-1"
  data-ouia-component-type="PF4/Toolbar"
  data-ouia-safe="true"
  data-testid="toolbar-header-testid"
  id="toolbar"
>
  <div
    class="pf-c-toolbar__content"
  >
    <div
      class="pf-c-toolbar__content-section"
    >
      <div
        class="pf-c-toolbar__item pf-u-mr-0"
        data-testid="filter-dropdown-testid"
      >
        <div
          class="pf-c-select"
          data-ouia-component-id="OUIA-Generated-Select-single-1"
          data-ouia-component-type="PF4/Select"
          data-ouia-safe="true"
          style="width: 11rem;"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Options menu"
            aria-labelledby=" pf-select-toggle-id-0"
            class="pf-c-select__toggle"
            id="pf-select-toggle-id-0"
            type="button"
          >
            <div
              class="pf-c-select__toggle-wrapper"
            >
              <span
                class="pf-c-select__toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                  />
                </svg>
              </span>
              <span
                class="pf-c-select__toggle-text"
              >
                Name
              </span>
            </div>
            <span
              class="pf-c-select__toggle-arrow"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="pf-c-toolbar__item"
        data-testid="filter-input-testid"
      >
        <div
          class="pf-c-input-group"
        >
          <div
            class="pf-c-search-input"
            id="textInput1"
            name="textInput1"
            type="search"
          >
            <div
              class="pf-c-input-group"
            >
              <div
                class="pf-c-search-input__bar"
              >
                <span
                  class="pf-c-search-input__text"
                >
                  <span
                    class="pf-c-search-input__icon"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Select input for Name"
                    class="pf-c-search-input__text-input"
                    placeholder="Filter by Name"
                    value=""
                  />
                </span>
                
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        align="[object Object]"
        class="pf-c-toolbar__item pf-m-pagination"
      >
        <div
          class="pf-c-skeleton"
          style="--pf-c-skeleton--Width: 200px;"
        >
          <span
            class="pf-u-screen-reader"
          />
        </div>
      </div>
    </div>
    <div
      class="pf-c-toolbar__expandable-content"
      id="toolbar-expandable-content-0"
    >
      <div
        class="pf-c-toolbar__group"
      />
    </div>
  </div>
  <div
    class="pf-c-toolbar__content"
  >
    <div
      class="pf-c-toolbar__content-section"
    >
      <div
        class="pf-c-toolbar__item pf-m-chip-group pf-m-spacer-none"
      />
    </div>
    <div
      class="pf-c-toolbar__expandable-content"
      id="toolbar-expandable-content-1"
    >
      <div
        class="pf-c-toolbar__group"
      />
    </div>
  </div>
  <div
    class="pf-c-toolbar__content pf-m-hidden"
    hidden=""
  >
    <div
      class="pf-c-toolbar__group"
    />
  </div>
</div>
`;
